### ISSUE TITLE
[Regression] -mss2 compiler flag on aarch64 build led to failure

### ISSUE THREAD

**Issue by luyatshimbalanga** [embree/embree#391]
Version 3.13.4 contains a regression where the configuration mechanism injects -mss2 flag on an aarch64 build as tested on Fedora build system resulting a failure. Version 4.13.3 is unaffected.

The result is attached on this 
[build.log.txt](https://github.com/embree/embree/files/9307120/build.log.txt)

**Comment by svenwoop** [comment]
Does the code compile properly when that -msse2 flags is remove? We have no testing of aarch64 under Linux setup at the moment.

**Comment by luyatshimbalanga** [comment]
In which files -msse2 is used? It seems `common/cmake/gnu.cmake` is insufficient.

**Comment by svenwoop** [comment]
That particular flag should just be set in that gnu.cmake file if you compile with GCC. What error message do you get?

**Comment by luyatshimbalanga** [comment]
This error 
```
cc1plus: error: unrecognized command-line option '-msse2'
```
Notice 
```
cd /builddir/build/BUILD/embree-3.13.4/common/math && /usr/bin/g++ -DEMBREE_TARGET_SSE2 -DTASKING_TBB  -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -mbranch-protection=standard -fasynchronous-unwind-tables -fstack-clash-protection -Wall -Wformat -Wformat-security -Wno-class-memaccess -fno-strict-overflow -fno-delete-null-pointer-checks -fwrapv -fsigned-char -fPIE -fPIC -std=c++11 -fvisibility=hidden -fvisibility-inlines-hidden -fno-strict-aliasing -fno-tree-vectorize -D_FORTIFY_SOURCE=2  -DNDEBUG -O3  -msse2 -std=gnu++17 -MD -MT common/math/CMakeFiles/math.dir/constants.cpp.o -MF CMakeFiles/math.dir/constants.cpp.o.d -o CMakeFiles/math.dir/constants.cpp.o -c /builddir/build/BUILD/embree-3.13.4/common/math/constants.cpp
```
Where embree sets -DEMBREE_TARGET_SSE2 even though NEON support is detected 
```
-- Detecting default ISA...
-- Detected default ISA: NEON
```

**Comment by luyatshimbalanga** [comment]
Looking at the [difference of commits](https://github.com/embree/embree/compare/v3.13.3...v3.13.4), it appears the conditional statement failed on the recently updated gnu.cmake,
```
_SET_IF_EMPTY(FLAGS_SSE2  "-msse2")
_SET_IF_EMPTY(FLAGS_SSE42 "-msse4.2")
_SET_IF_EMPTY(FLAGS_AVX   "-mavx")
_SET_IF_EMPTY(FLAGS_AVX2  "-mf16c -mavx2 -mfma -mlzcnt -mbmi -mbmi2")
_SET_IF_EMPTY(FLAGS_AVX512 "-mavx512f -mavx512dq -mavx512cd -mavx512bw -mavx512vl -mf16c -mavx2 -mfma -mlzcnt -mbmi -mbmi2 -mprefer-vector-width=256")
_SET_IF_EMPTY(FLAGS_NEON   "-D__SSE__ -D__SSE2__")
IF (EMBREE_ARM)
  SET(FLAGS_SSE2 "-D__SSE__ -D__SSE2__")
  SET(FLAGS_SSE42 "-D__SSE4_2__  -D__SSE4_1__")
  SET(FLAGS_AVX "-D__AVX__ -D__SSE4_2__  -D__SSE4_1__  -D__BMI__ -D__BMI2__ -D__LZCNT__")
  SET(FLAGS_AVX2 "-D__AVX2__ -D__AVX__ -D__SSE4_2__  -D__SSE4_1__  -D__BMI__ -D__BMI2__ -D__LZCNT__")
ELSE ()
  _SET_IF_EMPTY(FLAGS_SSE2  "-msse2")
  _SET_IF_EMPTY(FLAGS_SSE42 "-msse4.2")
  _SET_IF_EMPTY(FLAGS_AVX   "-mavx")
  _SET_IF_EMPTY(FLAGS_AVX2  "-mf16c -mavx2 -mfma -mlzcnt -mbmi -mbmi2")
  _SET_IF_EMPTY(FLAGS_AVX512 "-mavx512f -mavx512dq -mavx512cd -mavx512bw -mavx512vl -mf16c -mavx2 -mfma -mlzcnt -mbmi -mbmi2 -mprefer-vector-width=256")
ENDIF ()
```
It looks like the variable EMBREE_ARM failed to read NEON support leading to else statement hence assignment to "-msse2".

You can compare the build log from [embree 3.13.3 ](https://kojipkgs.fedoraproject.org//packages/embree/3.13.3/2.fc37/data/logs/aarch64/build.log)with the failed [3.13.4](https://kojipkgs.fedoraproject.org/work/tasks/6138/90716138/build.log)

**Comment by svenwoop** [comment]
We do not have a Linux aarch64 system in CI testing, thus we do not really support that target yet. Could you please help how to detect it? What does the CMAKE_SYSTEM_PROCESSOR variable contain on your system? Does the code compile if you enable EMBREE_ARM manually in cmake?

**Comment by gonzigonz** [comment]
I got around this with `cmake -DEMBREE_ARM=ON ..`.

**Comment by gonzigonz** [comment]
I can also confirm that in `v3.13.3` I can config, build and install with no issues but, in the latest release, I get the error above. The previous workaround I posted gets rid of the `-msse2` error but I can still not build, and instead get errors such as:
```
/simd/arm/sse2neon.h:6749:55: error: cannot convert ‘int64x2_t {aka __vector(2) long int}’ to ‘__m128i {aka __vector(4) int}’ in return
```

FWIW I am not on an Apple M1 machine, but a generic AArch64 CPU.

**Comment by luyatshimbalanga** [comment]
> We do not have a Linux aarch64 system in CI testing, thus we do not really support that target yet.
If you like, you can use the Fedora build system or [COPR](https://copr.fedorainfracloud.org/) which supports aarch64. You need to register as contributor for usage.

> Could you please help how to detect it? What does the CMAKE_SYSTEM_PROCESSOR variable contain on your system? 
I am not familiar with the use of that variable so I don't know how to properly use it in cmake.

Does the code compile if you enable EMBREE_ARM manually in cmake?
Unfortunately not as seem from the [build.log result](https://koji.fedoraproject.org/koji/taskinfo?taskID=91814753) as shared by @gonzigonz .

**Comment by Mluckydwyer** [comment]
I was able to compile without errorss by a adding the `-flax-vector-conversions` flag to the `common/cmake/gnu.cmake` file to allow for the vector conversions that cause errors. In the file it looks like this:

```CMake
...
IF (EMBREE_ARM)
  ...
  SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -flax-vector-conversions")  # allow lax vector type conversions
ENDIF (EMBREE_ARM)
...
```

This is after following @gonzigonz suggestion of:
> I got around this with cmake -DEMBREE_ARM=ON ...

This was all tested on a Nvidia Jetson AGX Xavier 16GB. I also was toying with adding support for detecting `aarch64` and enabling `EMBREE_ARM` by default and adding the flag, although setting it manually proved faster for now.

**Comment by svenwoop** [comment]
@Mluckydwyer What value is contained in the CMAKE_SYSTEM_PROCESSOR variable on your system?

**Comment by Mluckydwyer** [comment]
@svenwoop The value of `CMAKE_SYSTEM_PROCESSOR` is `aarch64` on my Nvidia Jetson running the latest version of Jetpack (5.0.2), which is a slimmed down Linux 64 bit OS based on Ubuntu 20.04 I believe.

I have also opened a PR with my changes that resolve this issue. It may be better to implement a separate `EMBREE_APPLE_ARM` variable to distinguish the Apple Silicon support from the generic ARM64 on Linux going forwards, but I wasn't sure the amount of internal CMake plumbing that it would take. Generic ARM64 is also not officially supported (as per your previous message), so this seemed more like a hot fix anyways. I also made no attempt to support ARM64 on Windows.
